### PR TITLE
If resize postmessage has been stringified, parse it

### PR DESCRIPTION
**Pull request by NataliaLKB**
A change suggested by @cramforce here:
https://github.com/ampproject/amphtml/issues/1408#issuecomment-171984571

Unfortunately my attention has been taken away from AMP for the next little while, so unsure when I can implement what was discussed in that issue. However, I can get interactives working on AMP pages at the Guardian with this PR.

Thanks for all your help :smile:

### CODE REVIEW

File: src/iframe-helper.js [hunk 1]
```diff
@@ -36,6 +36,9 @@ export function listen(iframe, typeOfMessage, callback, opt_is3P) {
   const sentinel = getSentinel_(opt_is3P);
   let unlisten;
   let listener = function(event) {
+    const data = typeof event.data == 'string'
+                ? JSON.parse(event.data)
```
Comment (cramforce) [inline src/iframe-helper.js]: Nit: indent only 4

File: src/iframe-helper.js [hunk 2]
```diff
@@ -36,6 +36,9 @@ export function listen(iframe, typeOfMessage, callback, opt_is3P) {
   const sentinel = getSentinel_(opt_is3P);
   let unlisten;
   let listener = function(event) {
+    const data = typeof event.data == 'string'
```
Comment (cramforce) [inline src/iframe-helper.js]: Please also check whether the first char is `{` and that `opt_is3P` is not true (This is performance critical for us).

File: src/iframe-helper.js (hunk 1 above)
Comment (cramforce) [inline src/iframe-helper.js]: Please file an issue to parse only once per received message (not once per listener).

File: src/iframe-helper.js (hunk 2 above)
Comment (NataliaLKB) [inline src/iframe-helper.js]: Yes that makes sense! I will add those checks right now

Comment (cramforce) [inline src/iframe-helper.js]: Oh, and you will need to wrap the JSON parse in a try block. Unfortunately we get a lot of crap messages and the `{` check might not be enough.

Comment (NataliaLKB) [inline src/iframe-helper.js]: Understandable. This is now updated

File: src/iframe-helper.js (hunk 1 above)
Comment (NataliaLKB) [inline src/iframe-helper.js]: https://github.com/ampproject/amphtml/issues/2836

File: src/iframe-helper.js [hunk 3]
```diff
@@ -36,6 +36,7 @@ export function listen(iframe, typeOfMessage, callback, opt_is3P) {
   const sentinel = getSentinel_(opt_is3P);
   let unlisten;
   let listener = function(event) {
+    const data = parseIfNeeded(event.data, opt_is3P);
```
Comment (jridgewell) [inline src/iframe-helper.js]: Can we move this down below the `event.source` conditional?

Comment (cramforce) [inline src/iframe-helper.js]: Great idea! That will save a lot of parsing.

**Comment by cramforce** [comment]
This is LGTM modulo @jridgewell's comment.

File: src/iframe-helper.js [hunk 4]
```diff
@@ -121,3 +122,25 @@ export function postMessage(iframe, type, object, targetOrigin, opt_is3P) {
...
+ */
+function parseIfNeeded(data, opt_is3P) {
+  const shouldBeParsed = typeof data === 'string'
+      && data.charAt(0) === '{'
+      && !opt_is3P;
+  if (shouldBeParsed) {
+    try {
+      data = JSON.parse(data);
+    }
+    catch (e) {
```
Comment (erwinmombay) [inline src/iframe-helper.js]: align catch to previous line please

`} catch (e) {`

**Comment by NataliaLKB** [comment]
Thank you @jridgewell @cramforce @erwinmombay. Changes have been made.

The travis build is failing on a test that passes locally, would one of you be able to restart it for me?

**Comment by cramforce** [comment]
Failure seems unrelated

**Comment by NataliaLKB** [comment]
Thank you @cramforce. Much appreciated

**Pull request merged**